jenkins/check_projects.py

```diff
@@ -0,0 +1,236 @@
+  sys.exit(1)
+
+
+def Sane(member):
+  email = member.split(':')[1]
+  return email.split('@')[0]
+
+
+def LoadProjects(configs):
+  projects = set()
+  for dirname, _, files in os.walk(os.path.dirname(__file__) or '.'):
+    for path in files:
+      if not path.endswith('.yaml'):
+        continue
+      with open(os.path.join(dirname, path)) as fp:
+        for project in re.findall(r'PROJECT="(.+)"', fp.read()):
+          if '{' not in project:
+            projects.add(project)
+            continue
+          elif '{version-infix}' not in project:
+            with lock:
+              print >>sys.stderr, 'project expansion not allowed ',
+              print project
+              errors.add(project)
+          for version in [
+              '1-0-1-2',
+              '1-1-1-2',
+              '1-1-1-3',
+              '1-2-1-3',
+              '1-2-1-4',
+              '1-3-1-4',
+          ]:
+            projects.add(project.replace('{version-infix}', version))
+  return projects
+
+
+def Update(project, role, member):
+  with semaphore:
+    err = subprocess.call(
+        [
+            'gcloud', '-q', 'projects',
+            'add-iam-policy-binding',
+            '--role=%s' % role,
+            '--member=%s' % member,
+            project,
+        ],
+        stdout=null,
+    )
+
+  if not err:
+    with lock:
+      print >>sys.stderr, 'Added %s as %s to %s' % (member, role, project)
+      updated.add(project)
+      return
+
+  with lock:
+    print >>sys.stderr, 'could not update ',
+    print project
+    errors.add(project)
+
+
+def Check(project, needed, mutate):
+  try:
+    with semaphore:
+      out = subprocess.check_output([
+          'gcloud',
+          'projects',
+          'get-iam-policy',
+          project,
+          '--format=json(bindings)',
+      ])
+  except subprocess.CalledProcessError:
+    with lock:
+      print >>sys.stderr, 'cannot access ',
+      print project
+      errors.add(project)
+      return
+
+  bindings = json.loads(out)
+  fixes = {}
+  for binding in bindings['bindings']:
+    role = binding['role']
+    members = binding['members']
+    if role in needed:
+      missing = set(needed[role]) - set(members)
+      if missing:
+        fixes[role] = missing
+    if role == 'roles/owner':
+      helpers[project] = members
+
+  if not fixes:
+    with lock:
+      print >>sys.stderr, project, 'already configured'
+      completed.add(project)
+      return
+
+  if not mutate:
+    with lock:
+      print >>sys.stderr, 'Will not --fix ',
+      print project
+      errors.add(project)
+      return
+
+  updates = []
+  for role, members in sorted(fixes.items()):
+    updates.extend(
+        threading.Thread(target=Update, args=(project, role, m))
+        for m in members
+    )
+
+  for update in updates:
+    update.start()
+
+  for update in updates:
+    update.join()
+
+
+if __name__ == '__main__':
+  CheckProjects(*sys.argv[1:])
```

Comment:
Strongly recommend using [argparse](https://docs.python.org/2/howto/argparse.html) instead of bare sys.argv. You'll get built in type checking and `--help` for free, and it's much easier to read and maintain.